[WIP] prototype: Add Avro reader

izchen opened the pull request:
Implemented a native Avro reader based on Apache avro-cpp (release-1.12.1).

Yuhta commented on velox/connectors/hive/HiveConnectorUtil.cpp:
```diff
@@ -604,6 +604,8 @@ void configureReaderOptions(
     readerOptions.setFileFormat(hiveSplit->fileFormat);
   }
+
+  readerOptions.serDeOptions().parameters = hiveSplit->serdeParameters;
```
We don't forward this blindly

izchen commented:
The `Ubuntu Benchmark` CI failed due to a false positive from static analysis, which is the same issue as reported in #15245. 
I plan to add `VELOX_SUPPRESS_STRINGOP_OVERFLOW_WARNING` to fix it.

izchen committed:
build: Add Avro reader (part 1/3): build and dependency setup

izchen committed:
feat: Add Avro reader (part 2/3): metadata and schema mapping

izchen committed:
feat: Add Avro reader (part 3/3): data decoding and vector write

izchen commented:
Hi @gggrace14 and @Yuhta, just following up here — wanted to make sure this didn’t get lost.

Whenever you have time, could you please take a look at the latest updates?

Happy to make further adjustments if needed. Thanks again!

Yuhta commented:
@izchen The reader schema is already provided with `RowReaderOptions::requestedType`, adding extra side way to pass it would cause duplication and inconsistency if they end up different.  Please construct the reader schema using `RowReaderOptions::requestedType` instead of custom property.

izchen committed:
feat: Respect requested type and scan spec in AvroReader

izchen commented:
> @izchen The reader schema is already provided with `RowReaderOptions::requestedType`, adding extra side way to pass it would cause duplication and inconsistency if they end up different. Please construct the reader schema using `RowReaderOptions::requestedType` instead of custom property.

@Yuhta Thanks for the comment, and sorry for the late reply.

I updated the behavior to make the Avro reader respect `RowReaderOptions::requestedType` and `scanSpec`:

- If `RowReaderOptions::requestedType` is provided, the Avro reader derives the target read schema from it, recursively aligns it with the file schema by field name, and validates the types. If it is not provided, the reader falls back to the file schema.
- If `RowReaderOptions::scanSpec` is provided, the Avro reader respects its projection, filter, constant, and mutation semantics, including pruning and filtering inside nested row / array / map fields.
- If both `requestedType` and `scanSpec` are provided, `requestedType` constrains the readable fields and types, while `scanSpec` determines the final output fields, ordering, filter-only fields, and constant fields.

I have also rebased the code onto the latest main.

As discussed earlier, I will split this large PR into smaller ones. The first one is here: https://github.com/facebookincubator/velox/pull/16555. Could you please take a look when you have a chance? I’m happy to make further adjustments as needed.

izchen commented:
@zhouyuan Thanks for the attention here.

Just a quick note that I have updated the PR. I’d appreciate it if you could also take a look when you have a chance.

zhouyuan commented:
@izchen thanks for the update, I ran into this gap when doing some iceberg test 
Cc: @rui-mo

pedroerp commented:
@izchen looks like this one is now waiting on your updates. Are you still interesting in finalizing this PR? Avro support would indeed be a nice feature to add to Velox.

Yuhta commented:
@pedroerp The actual merging should be broken down to smaller PRs, we can either close or put this PR in "Request changes" mode

jja725 commented on velox/dwio/avro/reader/AvroReader.cpp:
```diff
@@ -0,0 +1,1792 @@
+    if (pushback_ > 0) {
+      const void* rawData = nullptr;
+      int32_t size = 0;
+      stream_->Next(&rawData, &size);
```
what's the expected behavior if we reach EOF and return negative here.

jja725 commented on velox/dwio/avro/reader/AvroReader.h:
```diff
@@ -0,0 +1,93 @@
+  std::shared_ptr<AvroFileContents> contents_;
+};
+
+class AvroRowReader : public dwio::common::RowReader {
```
can we add some comments

jja725 commented on velox/dwio/avro/reader/AvroReader.cpp:
```diff
@@ -0,0 +1,1792 @@
+
+  vector_size_t numRead = 0;
+  while (numRead < rowsToRead) {
+    if (reader_->pastSync(splitLimit_) || !reader_->read(*datum_)) {
```
is there a plan to do vector reader

jja725 reviewed (changes requested):
Thanks for this great contribution and it's wanted within Uber as well! We would test it out internally so see if there's any gap/fix we can contribute as well. Also Agree we want seperate PR to make review easier.

izchen commented:
@pedroerp @Yuhta @jja725 Thanks for taking a look. I’ll update this PR later this week.

izchen committed:
Allow multiple row readers per file

izchen committed:
Correct reader behavior and build integration

izchen committed:
Correct schema and reader state handling

izchen committed:
Correct schema and reader state handling

izchen commented:
I reviewed and fixed a few correctness issues, and rebased the changes onto the latest main branch. I’ll split this PR into smaller ones shortly and create an issue to track them.
